diff --git a/pjsip/src/pjsip-simple/presence.c b/pjsip/src/pjsip-simple/presence.c
index 345eddd..1fd05fc 100644
--- a/pjsip/src/pjsip-simple/presence.c
+++ b/pjsip/src/pjsip-simple/presence.c
@@ -834,9 +834,14 @@ static void pres_on_evsub_rx_notify( pjsip_evsub *sub,
     if (rdata->msg_info.msg->body) {
 	status = pres_process_rx_notify( pres, rdata, p_st_code, p_st_text,
 					 res_hdr );
-	if (status != PJ_SUCCESS)
-	    return;
-
+        if (status != PJ_SUCCESS) {
+            /* Notify application even about bad notify that cannot be parsed. */
+            if (pres->user_cb.on_rx_notify) {
+                (*pres->user_cb.on_rx_notify)(sub, rdata, p_st_code, p_st_text,
+                                              res_hdr, p_body);
+            }
+            return;
+        }
     } else {
 #if 1
 	/* This is the newest change, http://trac.pjsip.org/repos/ticket/873
