diff --git a/libavformat/rtpenc.c b/libavformat/rtpenc.c
index 2ff59ee..b1940d3 100644
--- a/libavformat/rtpenc.c
+++ b/libavformat/rtpenc.c
@@ -48,52 +48,6 @@ static const AVClass rtp_muxer_class = {
 
 #define RTCP_SR_SIZE 28
 
-static int is_supported(enum AVCodecID id)
-{
-    switch(id) {
-    case AV_CODEC_ID_DIRAC:
-    case AV_CODEC_ID_H261:
-    case AV_CODEC_ID_H263:
-    case AV_CODEC_ID_H263P:
-    case AV_CODEC_ID_H264:
-    case AV_CODEC_ID_HEVC:
-    case AV_CODEC_ID_MPEG1VIDEO:
-    case AV_CODEC_ID_MPEG2VIDEO:
-    case AV_CODEC_ID_MPEG4:
-    case AV_CODEC_ID_AAC:
-    case AV_CODEC_ID_MP2:
-    case AV_CODEC_ID_MP3:
-    case AV_CODEC_ID_PCM_ALAW:
-    case AV_CODEC_ID_PCM_MULAW:
-    case AV_CODEC_ID_PCM_S8:
-    case AV_CODEC_ID_PCM_S16BE:
-    case AV_CODEC_ID_PCM_S16LE:
-    case AV_CODEC_ID_PCM_S24BE:
-    case AV_CODEC_ID_PCM_U16BE:
-    case AV_CODEC_ID_PCM_U16LE:
-    case AV_CODEC_ID_PCM_U8:
-    case AV_CODEC_ID_MPEG2TS:
-    case AV_CODEC_ID_AMR_NB:
-    case AV_CODEC_ID_AMR_WB:
-    case AV_CODEC_ID_VORBIS:
-    case AV_CODEC_ID_THEORA:
-    case AV_CODEC_ID_VP8:
-    case AV_CODEC_ID_VP9:
-    case AV_CODEC_ID_ADPCM_G722:
-    case AV_CODEC_ID_ADPCM_G726:
-    case AV_CODEC_ID_ADPCM_G726LE:
-    case AV_CODEC_ID_ILBC:
-    case AV_CODEC_ID_MJPEG:
-    case AV_CODEC_ID_SPEEX:
-    case AV_CODEC_ID_OPUS:
-    case AV_CODEC_ID_RAWVIDEO:
-    case AV_CODEC_ID_BITPACKED:
-        return 1;
-    default:
-        return 0;
-    }
-}
-
 static int rtp_write_header(AVFormatContext *s1)
 {
     RTPMuxContext *s = s1->priv_data;
@@ -105,11 +59,6 @@ static int rtp_write_header(AVFormatContext *s1)
         return AVERROR(EINVAL);
     }
     st = s1->streams[0];
-    if (!is_supported(st->codecpar->codec_id)) {
-        av_log(s1, AV_LOG_ERROR, "Unsupported codec %s\n", avcodec_get_name(st->codecpar->codec_id));
-
-        return -1;
-    }
 
     if (s->payload_type < 0) {
         /* Re-validate non-dynamic payload types */
