diff --git a/libavformat/rtpenc.c b/libavformat/rtpenc.c
index 38d931a..a72f95d 100644
--- a/libavformat/rtpenc.c
+++ b/libavformat/rtpenc.c
@@ -355,7 +355,7 @@ void ff_rtp_send_data(AVFormatContext *s1, const uint8_t *buf1, int len, int m)
 #else
     avio_w8(s1->pb, RTP_VERSION << 6);
 #endif
-    avio_w8(s1->pb, (s->payload_type & 0x7f) | ((m & 0x01) << 7));
+    avio_w8(s1->pb, (s->payload_type & 0x7f) | ((m & 0x00) << 7)); // never set marker for compatability reasons
     avio_wb16(s1->pb, s->seq);
     avio_wb32(s1->pb, s->timestamp);
     avio_wb32(s1->pb, s->ssrc);
